File: dotnet-service/src/Internal/Application/Services/CacheService.cs
using System;
using System.Text.Json;
using DotnetService.Internal.Config;
using DotnetService.Internal.Domain.Interfaces;
using DotnetService.Internal.Observability;

namespace DotnetService.Internal.Application.Services;

public enum CacheStatus
{
    Bypass,
    Miss,
    Fresh,
    Stale
}

public sealed record CacheResult<T>(T Value, CacheStatus Status);

public interface ICacheClient
{
    Task<string?> GetAsync(string key, CancellationToken cancellationToken);
    Task SetAsync(string key, string payload, TimeSpan ttl, CancellationToken cancellationToken);
    Task DeleteAsync(string key, CancellationToken cancellationToken);
}

public sealed class CacheService
{
    private  readonly  ICacheClient _client;
    private readonly ILoggerAdapter _logger;
    private readonly IClock _clock;
    private readonly bool _enabled;
    private readonly TimeSpan _defaultTtl;
    private readonly TimeSpan _staleTtl;

    public CacheService(ICacheClient client, CacheConfig config, ILoggerAdapter logger, IClock clock)
    {
        ArgumentNullException.ThrowIfNull(client);
        ArgumentNullException.ThrowIfNull(logger);
        ArgumentNullException.ThrowIfNull(clock);
        ArgumentNullException.ThrowIfNull(config);

        _client = client;
        _logger = logger;
        _clock = clock;
        _defaultTtl = config.DefaultTtl;
        _staleTtl = config.StaleTtl;
        _enabled = config.Enabled && _client is not null && _defaultTtl > TimeSpan.Zero;
    }

    public bool Enabled => _enabled;

    public async Task WriteAsync<T>(string key, T value, CancellationToken cancellationToken)
    {
        if (!_enabled)
        {
            return;
        }
        try
        {
            await PersistAsync(key, value, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            _logger.Warn("cache write failed", new { Key = key, Error = ex.Message });
        }
    }

    public async Task DeleteAsync(string key, CancellationToken cancellationToken)
    {
        if (!_enabled)
        {
            return;
        }
        try
        {
            await _client.DeleteAsync(key, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            _logger.Warn("cache delete failed", new { Key = key, Error = ex.Message });
        }
    }

    public async Task<CacheResult<T>> FetchAsync<T>(string key, Func<CancellationToken, Task<T>> loader, CancellationToken cancellationToken)
    {
        if (!_enabled || string.IsNullOrWhiteSpace(key))
        {
            var value = await loader(cancellationToken).ConfigureAwait(false);
            return Track(new CacheResult<T>(value, CacheStatus.Bypass));
        }

        try
        {
            var cached = await _client.GetAsync(key, cancellationToken).ConfigureAwait(false);
            if (!string.IsNullOrWhiteSpace(cached))
            {
                var envelope = JsonSerializer.Deserialize<CacheEnvelope<T>>(cached);
                if (envelope is not null)
                {
                    var age = _clock.UtcNow() - envelope.StoredAt;
                    if (age <= _defaultTtl)
                    {
                        return Track(new CacheResult<T>(envelope.Payload, CacheStatus.Fresh));
                    }

                    if (age <= _defaultTtl + _staleTtl)
                    {
                        _ = RefreshAsync(key, loader);
                        return Track(new CacheResult<T>(envelope.Payload, CacheStatus.Stale));
                    }
                }
            }
        }
        catch (Exception ex)
        {
            _logger.Warn("cache fetch failed", new { Key = key, Error = ex.Message });
            var fallback = await loader(cancellationToken).ConfigureAwait(false);
            return Track(new CacheResult<T>(fallback, CacheStatus.Bypass));
        }

        return await LoadAndStoreAsync(key, loader, cancellationToken).ConfigureAwait(false);
    }

    private async Task<CacheResult<T>> LoadAndStoreAsync<T>(string key, Func<CancellationToken, Task<T>> loader, CancellationToken cancellationToken)
    {
        var value = await loader(cancellationToken).ConfigureAwait(false);
        try
        {
            await PersistAsync(key, value, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            _logger.Warn("cache store failed", new { Key = key, Error = ex.Message });
        }
        return Track(new CacheResult<T>(value, CacheStatus.Miss));
    }

    private async Task PersistAsync<T>(string key, T value, CancellationToken cancellationToken)
    {
        if (!_enabled)
        {
            return;
        }

        var ttl = _defaultTtl + _staleTtl;
        if (ttl <= TimeSpan.Zero)
        {
            return;
        }

        var envelope = new CacheEnvelope<T>(value, _clock.UtcNow());
        var payload = JsonSerializer.Serialize(envelope);
        await _client.SetAsync(key, payload, ttl, cancellationToken).ConfigureAwait(false);
    }

    private Task RefreshAsync<T>(string key, Func<CancellationToken, Task<T>> loader)
    {
        return Task.Run(async () =>
        {
            try
            {
                var refreshed = await loader(CancellationToken.None).ConfigureAwait(false);
                await PersistAsync(key, refreshed, CancellationToken.None).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                _logger.Warn("cache refresh failed", new { Key = key, Error = ex.Message });
            }
        });
    }

    private static CacheResult<T> Track<T>(CacheResult<T> result)
    {
        MetricsRegistry.RecordCacheStatus(result.Status);
        return result;
    }

    private sealed record CacheEnvelope<T>(T Payload, DateTime StoredAt);
}

File: node-service/internal/application/services/cacheService.js
const metrics = require('../../observability/metrics');

const CacheStatuses = {
  BYPASS: 'bypass',
  MISS: 'miss',
  FRESH: 'fresh',
  STALE: 'stale'
};

class CacheService {
  constructor(client, config = {}, logger = console, clock = () => Date.now()) {
    this.client = client;
    this.logger = logger;
    this.clock = clock;
    this.enabled = Boolean(config.enabled) && Boolean(client);
    this.defaultTTL = Number(config.defaultTTL) || 0;
    this.staleTTL = Number(config.staleTTL) || 0;
  }

  isEnabled() {
    return this.enabled && this.defaultTTL > 0 && this.client;
  }

  async fetch(key, loader) {
    if (!this.isEnabled() || !key || typeof loader !== 'function') {
      const value = await loader();
      return this.track({ value, status: CacheStatuses.BYPASS });
    }

    try {
      const cached = await this.client.get(key);
      if (cached) {
        const parsed = JSON.parse(cached);
        const age = this.clock() - Number(parsed.stored_at || 0);
        if (age <= this.defaultTTL) {
          return this.track({ value: parsed.payload, status: CacheStatuses.FRESH });
        }
        if (age <= this.defaultTTL + this.staleTTL) {
          this.refresh(key, loader);
          return this.track({ value: parsed.payload, status: CacheStatuses.STALE });
        }
      }
    } catch (err) {
      this.logger?.warn?.('cache fetch failed', { key, error: err.message });
      const value = await loader();
      return this.track({ value, status: CacheStatuses.BYPASS });
    }

    return this.loadAndStore(key, loader);
  }

  async loadAndStore(key, loader) {
    const value = await loader();
    try {
      await this.store(key, value);
    } catch (err) {
      this.logger?.warn?.('cache store failed', { key, error: err.message });
    }
    return this.track({ value, status: CacheStatuses.MISS });
  }

  async store(key, value) {
    if (!this.isEnabled()) {
      return;
    }
    const ttl = this.defaultTTL + this.staleTTL;
    if (ttl <= 0) {
      return;
    }
    const entry = JSON.stringify({
      payload: value,
      stored_at: this.clock()
    });
    await this.client.set(key, entry, ttl);
  }

  async write(key, value) {
    try {
      await this.store(key, value);
    } catch (err) {
      this.logger?.warn?.('cache store failed', { key, error: err.message });
    }
  }

  async invalidate(key) {
    if (!this.isEnabled() || typeof this.client?.del !== 'function') {
      return;
    }
    try {
      await this.client.del(key);
    } catch (err) {
      this.logger?.warn?.('cache delete failed', { key, error: err.message });
    }
  }

  refresh(key, loader) {
    setImmediate(async () => {
      try {
        const value = await loader();
        await this.store(key, value);
      } catch (err) {
        this.logger?.warn?.('cache refresh failed', { key, error: err.message });
      }
    });
  }

  track(result) {
    metrics.recordCacheStatus(result.status);
    return result;
  }
}

module.exports = {
  CacheService,
  CacheStatuses
};
